pip_audit/_dependency_source/requirement.py

```diff
@@ -306,26 +297,30 @@ def _collect_cached_deps(
 
         new_cached_deps_for_file: dict[Requirement, set[Dependency]] = dict()
 
-        # There are three cases where we skip dependency resolution:
-        #
-        # 1. The user has explicitly specified `--require-hashes`.
-        # 2. One or more parsed requirements has hashes specified, enabling
-        #    hash checking for all requirements.
-        # 3. The user has explicitly specified `--no-deps`.
-        require_hashes = self._require_hashes or any(req.hash_options for req in reqs)
-        skip_deps = require_hashes or self._no_deps
-        if skip_deps:
-            for req, dep in self._collect_preresolved_deps(
-                iter(reqs), require_hashes=require_hashes
-            ):
+        # Skip dependency resolution if the user has specified `--no-deps`
+        if self._no_deps:
+            for req, dep in self._collect_preresolved_deps(iter(reqs)):
                 if req not in new_cached_deps_for_file:
                     new_cached_deps_for_file[req] = set()
                 new_cached_deps_for_file[req].add(dep)
                 yield req, dep
         else:
+            require_hashes = self._require_hashes or any(req.hash_options for req in reqs)
+            req_hashes = {}
+
+            # If we're requiring hashes, enforce that all requirements are hashed
+            if require_hashes:
+                for req in reqs:
+                    if not req.hash_options:
+                        raise RequirementSourceError(
+                            f"requirement {req.name} does not contain a hash {str(req)}"
+                        )
+                    assert req.name not in req_hashes
```

Comment:
Making sure I understand: what is this `assert` doing? It looks like it's checking to make sure we haven't seen this requirement before, but that's probably something that we should factor and ensure we check for all requirements files (given that it's a sign of self-inconsistency in the input).

Reply:
Yeah it's asserting that we haven't seen it before. We've already checked for duplicates elsewhere so this should never be the case.

I thought I'd just assert here so that if the code ever changes, we're going to fail loudly here instead of silently overwriting that dict key.

Reply:
Sure, I'll just turn that into a check and cover it with the unit tests.

Reply:
> The options are either to remove the assert and just keep it as is.

Yeah, let's go ahead and remove it then -- if we're already enforcing consistency elsewhere and we have unit tests backing that up, then I see no reason to dupe our work 🙂

Reply:
[0701244](https://github.com/pypa/pip-audit/pull/462/commits/0701244bf452d8917455ff2275db3daa3750c258)